os/kernel/sched/sched_migrate_tasks.c

```diff
@@ -0,0 +1,150 @@
+/****************************************************************************
+ * kernel/sched/sched_migrate_tasks.c
+ *
+ * Copyright 2025 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+
+#ifdef CONFIG_SCHED_MIGRATE
+
+#include <stdbool.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <tinyara/arch.h>
+#include <tinyara/spinlock.h>
+#include <tinyara/sched.h>
+
+#include "sched/sched.h"
+#include "clock/clock.h"
+
+
+/****************************************************************************
+ * Name: sched_migrate_tasks
+ *
+ * Description:
+ *   Migrates all tasks from a specified offline CPU to other available
+ *   online CPUs.
+ *
+ *   IMPORTANT: This function must be called from a running CPU (e.g., CPU0)
+ *   The caller must ensure preemption is locked or this is called
+ *   in a context where it cannot be preempted.
+ *
+ * Inputs:
+ *   offline_cpu - The index of the CPU that needs to be shut down.
+ *
+ * Return Value:
+ *   OK (0) on success, or a negated errno value on failure.
+ *
+ ****************************************************************************/
+
+int sched_migrate_tasks(int offline_cpu)
+{
+	FAR struct tcb_s *tcb;
+	FAR struct tcb_s *priv_tcb;
+	FAR dq_queue_t *task_list;
+	irqstate_t flags;
+	int ret = OK;
+
+	/* Basic sanity checks */
+	if (offline_cpu < 0 || offline_cpu >= CONFIG_SMP_NCPUS) {
+		slldbg("ERROR: Invalid offline_cpu: %d\n", offline_cpu);
+		return -EINVAL;
+	}
+
+	if (offline_cpu == this_cpu()) {
+		slldbg("ERROR: Cannot migrate tasks from the current running CPU: %d\n", offline_cpu);
+		return -EINVAL;
+	}
+
+	/* Lock the scheduler to prevent preemption during this operation.
+	 * This ensures that no other tasks can be scheduled while we're
+	 * migrating tasks from the offline CPU.
+	 */
+	sched_lock();
+
+	/* Disable interrupts to ensure atomicity throughout this critical operation. */
+	flags = enter_critical_section();
+
+	/* Update the active CPUs mask to indicate that the offline CPU is no longer active */
+	CPU_CLR(offline_cpu, &g_active_cpus_mask);
+
+	DEBUGVERIFY(up_cpu_pause(offline_cpu));
+
```

Comment:
you can remove it here. pause/resume

sched_removereadytorun is already having it.